---
title: "Power Analysis of Microbiome Data"
author: "Michael Agronah"
#date: "2022-10-27"
date: '\today'
#date: '`r format(Sys.time(), "%d %B, %Y")`'
output:
  pdf_document:
    toc: true    # can remove
    toc_depth: 3 # can remove
    number_sections: true
  highlight: zenburn
header-includes:
- \usepackage{fancyhdr}
- \usepackage{mathtools}
- \usepackage{xcolor, hyperref}
- \usepackage{lipsum}
- \usepackage{caption}
- \usepackage{diagbox}
- \usepackage{multirow}
- \usepackage{amsmath}
- \setlength{\headheight}{28pt}
- \setlength{\footskip}{25pt}
- \pagestyle{fancy}
- \renewcommand{\headrulewidth}{0.5pt}
- \renewcommand{\footrulewidth}{0.5pt}
- \lhead{\includegraphics[width=8cm,height=1cm]{logo-McMaster}}
- \cfoot{School of Computational Science and Engineering \\ McMaster University}
- \rhead{\thepage}
- \hypersetup{colorlinks   = true, linkcolor=blue, urlcolor  = blue}
- \fancypagestyle{plain}{\pagestyle{fancy}}
editor_options: 
  chunk_output_type: console
bibliography: packages.bib
---

```{r load_pkgs, include=FALSE, message=FALSE}
library(ggplot2)
library(fitdistrplus)
library(truncnorm)
library(BiocManager)
if (FALSE) BiocManager::install(c("DESeq2", "edgeR"))
library(nlme)
library(sn)
library(truncdist)
library(DESeq2)
library(tidyverse)
library(bbmle)
library(patchwork)
library(truncdist)
library(plyr)
library(metR)  
library(mgcv)
library(ggplot2)
library(patchwork)
library(extraDistr)
library(here)
setwd(here())
theme_set(theme_bw())
```
\section*{Summary}
This document constitute 2 parts. Part 1 is a project the investigates the relationship between power, effect sizes and abundances in differential microbiome abundance studies. Part 2 focuses on investigating the modeling of the trends in longitudinal microbiome data and the relationship between longitudinal microbiome data with variables of interest such as the disease status of patients, and environmental factors. 

# Past 1 Power Analysis of Microbiome Data
## Abstract 
Microbiome studies are generally underpowered. Consequently, significance results reported in the literature can be misleading. Furthermore, the dynamics underlying the relationships among effect sizes, power and abundances in differential abundance studies is unclear. The ability to predict, even before the commencement of a research, how many species a researcher can reliably detect power for is an important question in differential abundance studies. This project seeks to develop a paradigm for understanding the relationships among effect sizes, power and abundances as well as the underpinning mechanism driving these relationships. 

We investigate these relationships using ten (10) 16S microbiome datasets from children with autism spectrum disorder (ASD) (as treatment group) and neurotypical children (as control group).  Effect size in this study is defined by log2 fold changes. Control is used to refer to the log2 control abundances.

Current results from the project show a positive relationship between standardized effect size and power.  An R package will be developed from this project where a new dataset can be fed into this package to explore the power,  effect and abundance relationships. 

<!-- The findings demonstrated that the effect size distributions followed a cauchy distribution, which defines the variance defined as a quadratic function of control abundances. -->

# Introduction 
A key question in differential abundance studies is to determine differences in microbiome abundance between various groups of interest. For instance, what are the differences between microbiome abundances in control and treatment groups? This question can be translated into a hypothesis testing procedure. The standard process in hypothesis testing is to first define a null hypothesis, choose a significant level and reject the null hypothesis whenever p-value is below the significant level or fail to reject the null hypothesis whenever p-value is above the significant level. This technique has two potential drawbacks:  first, failing to reject the null hypothesis when the p-value is greater than the null hypothesis merely by accident (resulting in what is known as the Type 1 error or the false positive error) and second, rejecting the null hypothesis when the p-value is less than the significant level by chance (known as Type 2 error false negative error). 
<!-- These errors are illustrated in figure \ref{}. -->

Hypothesis testing depends on 4 parameters: (i) the effect size, which is a measure of the magnitude of differences between groups of interest; (ii) the sample size $(n)$, the number of samples to be used in the studies; (iii) the power of a study ($1 - \beta$), the probability of correctly rejecting the null hypothesis; and (iv) the confidence level ($\alpha$), the probability of getting a false positive, that is, rejecting the null hypothesis will when it is actually true. In general, power, sample sizes, significance level and effect size are all positively related. For example, increasing the significance level increases power and a higher effect size will generally lead to higher power [@xia2018statistical]. Power studies are important to minimize type 1 and type 2 errors [@xia2018statistical] and to determine the success of a study design even before commencing. In many funding applications, power analysis on a simulation study is required to evaluate how many sample sizes are needed to detect what effect sizes and the power of the studies.

<!-- It also helps to see whether a study will be successful before you even begin the studies.  -->


<!-- [(#fig:bar) Illustration of Type 1 and Type 2 errors. Image source: [Graduatetutor](https://www.graduatetutor.com/statistics-tutor/type-1-type-2-errors-hypothesis-testing-statistics).](error.png) -->

The majority of microbiome studies are underpowered [@kers2021power; @brussow2020problems]. Consequently, significance results reported in the literature of differential abundance microbiome studies are influenced by type 1 and type 2 errors. Failure to correct for these errors has resulted in numerous instances of conflicting conclusions in the literature of microbiome studies [@brussow2020problems]. Early research on the relationship between the gut microbiota of obese and non-obesity groups, for example, found significant differences in the diversity of the gut microbiota and significant differences in the Bacteroidetes/Firmicutes (B/F) ratio between obese and non-obese people [@ley2006human]. Later research, however, using larger samples, found only slight differences in microbial diversity and no statistically significant change in the B/F ratio between non-obese and obese individuals [@sze2016looking]. This generally makes it challenging to reproduce findings from microbiome studies reported in literature [@kers2021power]. Due to the failure to correct type1 and type 2 errors, an average effect size and power of studies in the literature of differential abundance microbiome studies is unknown. This makes is difficult to make informed decisions for conducting a power studies of a design prior to commencing a study. 

<!-- One potential remedy is to gain a better understanding of the relationships between power effect size and abundances in differential abundance microbiome studies as well as the mechanisms underlying these relationships. This project seeks to develop a framework for understanding theses relationships among effect sizes, power and abundances as well as the underpinning mechanism driving these relationships.  -->

One potential remedy is to gain a better understanding of the relationships between power, effect size and abundances in differential abundance microbiome studies. This project seeks to develop a paradigm for understanding the relationships among effect sizes, power and abundances as well as the underpinning mechanism driving these relationships.  Using this framework from this project new dataset can be to explored to determine the relationships between power,effect size and abundance in the context of differential abundance studies. 

<!-- One possible solution is to conduct a power analysis before beginning the studies. -->
<!-- The typical effect sizes to use when conducting power analysis, on the other hand, are unknown. The ability to determine how many samples are required to detect a certain level of power given an effect size is of great interest in microbiome studies. -->
<!-- Also, before beginning a study, one would typically like to know how many taxa they can reliably detect power for. -->


## Research goals 

#### Project Contributions:

This project contributes to the current literature on power analysis in microbiome studies in the following ways

1. develop a framework for estimating the power and effect sizes at the taxon level and describing the relationship between effect sizes and power and abundances 
for differential abundance studies that use the Negative Binomial Model. Two commonly used packages that implement the Negative Binomial Model are Deseq2 and edgeR. For simplicity, this study focuses on the implementation in the Deseq2 package. However, the same analysis can be repeated using the edgeR package as well. This will help researchers to bla bla bla

2. develop and R package with visualization tools for power calculations and effect size at these. This will help researchers to bla bla bla…

The project will answer the following questions. 

1. What is the relationship between effect size, power and abundances? For example, what is the average power and effect size for abundances in the range. 

2.How many sample sizes ste needed to detect a given power and at a given effect size.

3. For a given effect size will determine the number of taxa with power greater than a desire value

#### Challenges:

To do power analysis, one needs to know the effect size and the sample size. A typical approach is to get the effect sizes reported in the literature for the power. The accuracy  of these effect sizes, however, is unclear. Often these effect sizes of microbiome studies are not reported [@ ]. …showed that the effects from these studies are biased . 

Mm reviewed…papers . In this studies, 100 publications reviewed did not report  effect sizes from the studies 

Therefore it is unclear what an average effect size for differential abundance analysis in microbiome research is. 

Power is more likely overestimated in differential abundance studies in the literature and the relationship between power, effect size and abundances is unknown. 

5o the nest of pur knowledge,  no studies exist that explores the effect sizes and power for different ranges of abundances. 

<!-- The goal of this project is to  -->

<!-- 1. investigate the relationships between power, effect sizes and abundances with the aim of understanding the mechanisms behind these relationships.  -->

<!-- 2. investigate the number of individual taxa in a differential abundance microbiome -->
<!-- studies that power can be detected reliably, for a given effect and sample size. -->

<!-- 3. investigate the number of sample sizes needed to detect a given power at various -->
<!-- effect sizes -->




## Research Design and Method

### Data collection and processing 
Using the search terms *"autism[All Fields] AND 16S[All Fields]"*, *"autism[All Fields] AND 16S[All Fields] AND Fecal[All Fields]"*, raw sequence data from 10 projects that examined the microbiome of children with autism spectrum disorder were gathered from the European Nucleotide Archive EBA and the National Center for Biotechnology Information NCBI. The following are the accession numbers for the 10 projects in the NCBI and ENA archives: PRJNA687773, PRJNA624252, PRJNA453621, PRJNA642975, PRJNA355023, PRJNA168470, PRJNA644763, PRJNA578223, PRJNA589343, and PRJEB45948. Children with autism spectrum disorders represent the treatment group in these datasets, whereas children with neurotypical or typical development are the control groups. Adaptor and primer sequences were removed using "cutadapt" function implemented in a bash script. The trimmed sequence from the "cutadpt" were then processed into Amplicon Sequence variants (ASVs) data using the Dada2 pineline which involves filtering and trimming, error estimation, denoising, merging paired reads and chimeras removal [@chen2020gut]. 

### Models description
The null hypothesis ($H_0$) and the alternative hypothesis ($H_a$) for testing differences between 2 groups (control and treatment) is formulated by  $H_0: \mu_{control} - \mu_{treatment} = 0$ and $H_a: \mu_{control} - \mu_{treatment} \neq$ respectively, where $\mu_{control}$ and $\mu_{treatment}$ denote the means counts in control and treatment groups respectively. Effect sizes,a measure of the tdifference in the mean can then be tested.  The negative binomial distribution has often used to model microbiome count data due to the presence of overdispersion in microbiome count data. Let $K_ij$ denote the count data for the $i^{th}$ taxa in the $j^{th}$ sample.  $K_ij$  is modeled by  negative binomial distribution defined  by
\begin{align}
K_{ij} \sim NB(mean &= \mu_{ij}, dispersion = \alpha_i), \\
\mu_{ij} &= g^{-1}(E_{ij})\\ 
E_{ij} &= \sum_r x_{jr}\beta_{ir} ,
\end{align}
where $g$ is a link function, $\beta_{ir}$ are estimates of the effect sizes and $x_{jr}$ are the covariates. The relationship between the variance of counts and the dispersion is given by $Var K_{ij} = \mu + \alpha\mu^2$. In this project, the estimating procedure implemented in the *DESeq2* package in R is used for estimating $\beta_{ir},\mu_{ij}$ and $\alpha_i$. Details concerning this procedure is stated in the paper by @love2014moderated and by @anders2010differential. 

```{r load_data, include=FALSE, message=FALSE, echo=FALSE}
path <- "Power_Project/Datasets/"
plot_path <- "Latex_Files/Proposal/Figs/power/"
source("Power_Project/R_files/Functions.R")

plotname <- "box_plot_effects_"
#########################################
data_list <- mget(load("Power_Project/Datasets/data.RData"))
metadata_list <- mget(load("Power_Project/Datasets/groups.RData"))
name <- names(metadata_list); stopifnot(names(data_list) == name)
############################################################
subset <- c("PRJNA453621","PRJEB45948","PRJNA589343", "PRJNA687773")
names(subset) <- subset
data_list <- list(data_list[["PRJNA453621"]], data_list[["PRJEB45948"]],
                  data_list[["PRJNA589343"]], data_list[["PRJNA687773"]])

metadata_list <- list(metadata_list[["PRJNA453621"]],
                      metadata_list[["PRJEB45948"]],
                      metadata_list[["PRJNA589343"]],
                      metadata_list[["PRJNA687773"]])

names(data_list) <- names(subset)
names(metadata_list) <- names(subset)
name = names(subset)
######################################################################
DeseqRes <- mget(load(paste0(path,"DeseqResults.RData"))) #list of deseq results
deseq_list <- DeseqRes[[names(DeseqRes)]]   #extra Deseq data from the list
stopifnot(names(deseq_list) == name)

Control_Treat <- mget(load(paste0(path,"Control_Treat_Data.RData"))) #list of deseq results
Control_Treat <- Control_Treat[[names(Control_Treat)]]
stopifnot(names(Control_Treat) == name)

param.disp <- mget(load(paste0(path,"Dispersion_Parameters.RData"))) 
param.disp <- param.disp[[names(param.disp)]]

logcontrol <- read_data(Control_Treat, "LogControl") #extract control mean datasets from the list
```

### Modelling the relationship between control abundances and effect sizes 
Plots of the control abundances and effect sizes are shown in figure \ref{cont_eff}. The plots display an average effect size of zero. According to the plot, variations around the smooth curve are greatest in the middle and lowest at the ends of the smooth curves. A quadratic function can be used to describe the variance of the effect sizes as a function of control abundances. A scale location plot shown in figure \ref{scale_loc}  validates the quadratic relationship for the variance. A skewed normal distribution and a truncated Cauchy distribution, respectively, can be used to approximate the distribution of control abundances and effect sizes. Figure \ref{fit_con} and figure \ref{fit_eff} show a histogram and density of the control abundance and effect sizes, as well as a plot of the density from simulation from fitting these distributions to the data. Effect sizes were fitted to Cauchy distributions defined with zero location and scale parameters defined by a quadratic function of control abundance, as shown in equation by \ref{eq1}.
\begin{align} \label{eq1}
y_i &= Cauchy(location = 0, scale = \gamma_i), \gamma_i =  \exp(k_0 +  k_1x_i + k_2{x_i}^2),
\end{align}
where $y_i$ and $x_i$ are the effect size and control abundance, respectively, for the $i^{th}$ taxa and $k_j ; j = 1, 2, 3$ are constants to be estimated. 

```{r control_effects, echo=FALSE, warning=FALSE, message=FALSE, out.width="60%",fig.align="center", fig.cap="\\label{scale_loc}  Plot of control against effect sizes for datasets with assession numbers PRJNA453621, PRJEB45948, PRJNA589343 and PRJNA687773", fig.height = 5}
plts=Plot_ContEffects(deseq_list,logcontrol, plot_path)
m = plts[[1]]; n = plts[[2]]
(m[[1]]|m[[2]])/(m[[3]]|m[[4]])
```

```{r scale_location plot, echo=FALSE, warning=FALSE, message=FALSE, out.width="60%",fig.align="center", fig.cap="\\label{cont_eff} Scale-location plot of control against squared fold changes for datasets with assession numbers PRJNA453621, PRJEB45948, PRJNA589343 and PRJNA687773", fig.height = 5}
(n[[1]]|n[[2]])/(n[[3]]|n[[4]])
```

```{r control_fits, warning=FALSE, message=FALSE, echo=FALSE,out.width="70%",fig.align="center", fig.cap="\\label{fit_con} Histogram and density plots for control abundances and density from a simulated sample from fitting a truncated  normal distribution"}
file_path <- "control_histograms_"
v1 = Fit_control(logcontrol,plot_path,file_path, plotname)
v1 = (v1[[1]]|v1[[2]])/(v1[[3]]|v1[[4]]) & theme(legend.position = "bottom")
v1 + plot_layout(guides = "collect")
``` 


```{r effect_fits, warning=FALSE, message=FALSE, echo=FALSE,out.width="70%",fig.align="center", fig.cap="\\label{fit_eff} Histogram and density plots for control abundances and density from a simulated sample from fitting a Cauchy distribution to the data"}
param.control <- mget(load(paste0(path,"Control_Parameters.RData"))) #lis lts
param.control <- param.control[[names(param.control)]]

effects_list <- mget(load(paste0(path,"DeseqFoldChanges.RData"))) 
effects_list <- effects_list[[names(effects_list)]]

plotname  <- "effects_dist_"
v = Fit_Effects(effects_list,logcontrol,plot_path,plotname)
vv = (v[[1]]|v[[2]])/(v[[3]]|v[[4]]) & theme(legend.position = "bottom")
vv + plot_layout(guides = "collect")
```


### Estimation of standard errors
### The delta method
Write up on the delta method. The delta method will be used to estimate the standardised error for the control abundance and the effect sizes defined

The delta method is a method for claucluationg the  assymptotic bal bal bale.
For a random variable X and parameter $\theta, \hat{\theta}$ and any smooth function f   with the property that $f'(\theta)$ exists and is non-zero valued, the delta method states that
$$
\frac{\hat{\theta} - \theta}{\hat{SE}_\theta} \xrightarrow{} N(0,1),\,\, then \\
\frac{f(\hat{\theta}) - f(\theta)}{f'(\theta) \hat{SE}_\theta} \xrightarrow{} N(0,1)
$$
and tha asymptotic mean and asymptotic standard error of $f(\hat{\theta})$ is $f(\theta)$ and $f('{\hat{\theta}}) \hat{SE}(\theta)$ respectively.


Let $C$ and $T$ denote the control and treatment groups respectively. The standard error of  fold changes is define be estimated as follows:

$$\textbf{var}[\log_2\left(\frac{T}{C}\right)] = \textbf{var}[\log_2(T)] + \textbf{var}[\log_2(C)]$$
From Taylor series expansion $\textbf{var}[f(X)]  = \textbf{var}[f(\mu_X + X - \mu_X)]  \approx \textbf{var}[f(\mu_X) + f'(\mu_X)(X - \mu_X) + \frac{f''(\mu_X)(X - \mu_X)]}{2} + ..= f'(\mathrm{\mathbb{E}}[X])^2\textbf{var}[X] + \frac{f'(\mathrm{\mathbb{E}}[X])^2\textbf{var}[X]}{4}$
,

$$
\begin{aligned}
for f(X) = \log_2(X), f' &= \frac{1}{x\ln 2}\\
\textbf{var}[\log_2\left(\frac{T}{C}\right)] &= \textbf{var}[\log_2(T)] + \textbf{var}[\log_2(C)]\\
&\approx \frac{1}{T\ln 2}\textbf{var}[T]    +  \frac{1}{C\ln 2}\textbf{var}[C]\\
\mathbb{SE}[\log_2\left(\frac{T}{C}\right)] &= \sqrt{\frac{1}{T\ln 2}\textbf{var}[T]    +  \frac{1}{C\ln 2}\textbf{var}[C]}
\end{aligned}
$$



```{r heatmap_plot,warning=FALSE, message=FALSE, echo=FALSE}
#################################################################
n_otu <- 300; n_samples <-  20; n_sim = 100
param.effects <- mget(load(paste0(path,"Effect_Parameters.RData")))
param.effects <- param.effects[[names(param.effects)]]
#data_sim(param.effects,param.disp,param.cont=param.control,n_sim,n_samples,n_otu,path,filter=10)
sim_data_list <- mget(load(paste0(path,"Simulated_Data.RData")))
sim_metadata_list <- mget(load(paste0(path,"Simulated_Metadata.RData")))
true_control_list <- mget(load(paste0(path,"True_Control.RData")))
true_effect_list <- mget(load(paste0(path,"True_Effect.RData")))
standard_err_delta_list <- mget(load(paste0(path,"Standard_Err_Delta.RData")))
#View(true_effect_list)
standard_err_delta_list = standard_err_delta_list[[names(standard_err_delta_list)]]

sim_data_list = sim_data_list[[names(sim_data_list)]]
sim_metadata_list = sim_metadata_list[[names(sim_metadata_list)]]
true_control_list = true_control_list[[names(true_control_list)]]
true_effect_list = true_effect_list[[names(true_effect_list)]]
###############################################################
#Deseq1(sim_data_list,sim_metadata_list,path)
deseq_data_list <- mget(load(paste0(path,"Sim_Deseq_RESULT.RData")))
deseq_data_list <- deseq_data_list[[names(deseq_data_list)]]
View(deseq_data_list[[1]][[1]])

#max(deseq_data_list[[1]][[1]]$lfcSE)
#View(combined_data[[1]]$deltaStandard_error)
#########################################################
combined_data =combined_data_list(true_control_list, true_effect_list,
                                 deseq_data_list,standard_err_delta_list, alpha=0.1)
r = Power_Heatmap(combined_data); plts1 = r[["full_plot"]]; plts2 = r[["reduced_plot2"]]

#b = (plts1[[1]]|plts1[[2]]) & theme(legend.position = "bottom")
#b + plot_layout(guides = "collect")

#b =(plts1[[3]]|plts1[[4]])  & theme(legend.position = "bottom")
#b + plot_layout(guides = "collect")
```


<!-- \newpage -->
<!-- ## Result -->
<!-- This section presents the current findings from this projects.  -->
<!-- We simulate 200 datasets with 1000 and 100 samples each in the control and treatment groups for each of the datasets. Figure \ref{fit_heat1} and figure \ref{fit_heat2} displays heatmaps of the power and total control abundances for different gids of control abundances and effect sizes. In each plot, regions with total control abundances less than 100 have been filtered out. -->

<!-- <!-- The plot to the left of figure ... for instance shows an average the power of about The plots sspread of power at various effect sizes for each of the datasets. Th --> -->

<!-- Figure \ref{fit_heat1} and figure \ref{fit_heat2} shows the contour plots for powers. From the plots, regions without data should be ignore -->
<!-- THe delta method is used to approximate the standard errors of the effect size estimates. Figure and figure .. shows the plots for the power as contours when the standardised effect size is used.  -->

<!-- <!-- Form the first plot, regions with high concentrations of this pattern may be --> -->
<!-- ```{r heatmap_1, warning=FALSE, message=FALSE, echo=FALSE,fig.cap="\\label{fit_heat1} Heatmap showing the power and total number of taxa in grids of control and effect sizes"} -->
<!-- library(png) -->
<!-- library(grid) -->
<!-- img <- readPNG("Power_Project/Datasets/heatmapnew1_png") -->
<!-- grid.raster(img) -->
<!-- ``` -->

```{r contour, warning=FALSE, message=FALSE, echo=FALSE,fig.width=14, fig.height=12,fig.cap="\\label{fit_heat1} Heatmap showing the power and total number of taxa in grids of control and effect sizes"}
combined_data <- mget(load("Power_Project/Datasets/combined_data_list.RData"))
combined_data <- combined_data[[names(combined_data)]]
r = Power_Heatmap(combined_data)
plts1 = r[["full_plot"]]; plts2 = r[["reduced_plot2"]]

pr =contour_plots(combined_data, standardised = TRUE,
                  interval=c(0.01,0.01,0.01,0.01))
                  #interval=c(0.005,0.0005,0.005,0.005))

n = (pr[[1]][["contour_plot"]]|pr[[2]][["contour_plot"]])/(pr[[3]][["contour_plot"]]|pr[[4]][["contour_plot"]]) & theme(legend.position = "bottom")
n + plot_layout(guides = "collect")
```


<!-- ```{r heatmap_2, warning=FALSE, message=FALSE, echo=FALSE,out.width="100%",fig.cap="\\label{fit_heat2} Heatmap showing the power and total number of taxa in grids of control and effect sizes"} -->
<!-- library(png) -->
<!-- library(grid) -->
<!-- img <- readPNG("Power_Project/Datasets/heatmapnew2_png") -->
<!-- grid.raster(img) -->
<!-- ``` -->


<!-- ```{r contour_1, warning=FALSE, message=FALSE, echo=FALSE,out.width="100%",fig.cap="\\label{fit_heat2} Heatmap showing the power and total number of taxa in grids of control and effect sizes"} -->
<!-- library(png) -->
<!-- library(grid) -->
<!-- img <- readPNG("Power_Project/Datasets/contour1_png") -->
<!-- grid.raster(img) -->
<!-- ``` -->

<!-- #Conclusion  -->

<!-- # Future work -->

<!-- 1. Future work, we seek to explore the relationship between standard error and how standard error is dringving the power of otus  -->
<!-- 2. Intuitively, it seem power should be independent of  -->
<!-- control abundances when we are looking at the standardised effect size. So we will included higher approximations for the for the computation of the standard error.  -->
<!-- 3. We would like to see how the standard error is influencing power. So we will compute standard error more and show the plot of it. The standard error itself should be comming from the combination of errors in the fit for the control effect and the varaince in the negeative binomial samples. So we see to understand that.  -->




<!-- ```{r heatmap_contour1, warning=FALSE, message=FALSE, echo=FALSE,fig.cap="\\label{fit_cont} Heatmap showing the power and total number of taxa in grids of control and effect sizes"} -->
<!-- library(png) -->
<!-- library(grid) -->
<!-- img <- readPNG("Power_Project/Datasets/heatmap_contour1_png") -->
<!-- grid.raster(img) -->
<!-- ``` -->


<!-- ```{r heatmap_contour2, warning=FALSE, message=FALSE, echo=FALSE,fig.cap="\\label{fit_cont} Heatmap showing the power and total number of taxa in grids of control and effect sizes"} -->
<!-- library(png) -->
<!-- library(grid) -->
<!-- img <- readPNG("Power_Project/Datasets/heatmap_contour2_png") -->
<!-- grid.raster(img) -->
<!-- ``` -->

<!-- ```{r heatmap_contour3, warning=FALSE, message=FALSE, echo=FALSE,fig.cap="\\label{fit_cont} Heatmap showing the power and total number of taxa in grids of control and effect sizes"} -->
<!-- library(png) -->
<!-- library(grid) -->
<!-- img <- readPNG("Power_Project/Datasets/heatmap_contour3_png") -->
<!-- grid.raster(img) -->
<!-- ``` -->


<!-- ```{r heatmap_contour4, warning=FALSE, message=FALSE, echo=FALSE,fig.cap="\\label{fit_cont} Heatmap showing the power and total number of taxa in grids of control and effect sizes"} -->
<!-- library(png) -->
<!-- library(grid) -->
<!-- img <- readPNG("Power_Project/Datasets/heatmap_contour4_png") -->
<!-- grid.raster(img) -->
<!-- ``` -->


<!-- ```{r heatmapplot1, warning=FALSE, message=FALSE, echo=FALSE,fig.show="hold",fig.cap="\\label{fit_cont} Heatmap showing the power and total number of taxa in grids of control and effect sizes"} -->
<!-- library(png) -->
<!-- library(grid) -->
<!-- img1 <- readPNG("Power_Project/Datasets/heatmap1_png") -->
<!-- grid.raster(img1) -->
<!-- img2 <- readPNG("Power_Project/Datasets/heatmap2_png") -->
<!-- grid.raster(img2) -->
<!-- ``` -->

<!-- ```{r heatmapplot2, warning=FALSE, message=FALSE, echo=FALSE,fig.cap="\\label{fit_cont} Heatmap showing the power and total number of taxa in grids of control and effect sizes"} -->
<!-- library(png) -->
<!-- library(grid) -->
<!-- img2 <- readPNG("Power_Project/Datasets/heatmap2_png") -->
<!-- grid.raster(img2) -->
<!-- ``` -->



<!-- ```{r heatmap2, include=FALSE, warning=FALSE, message=FALSE, echo=FALSE,fig.cap="\\label{fit_cont} Heatmap showing the power and total number of taxa in grids of control and effect sizes"} -->
<!-- #b2 = (plts2[[1]]|plts2[[2]])/(plts2[[3]]|plts2[[4]]) & theme(legend.position = #"bottom") -->
<!-- #b2 + plot_layout(guides = "collect") -->
<!-- #standard_lfc = true_effect/sd_error -->
<!-- ########################################################################### -->
<!-- ``` -->



<!-- ```{r image, echo=FALSE, fig.cap="\\label{fig:error}", out.width = '100%'} -->
<!-- #knitr::include_graphics(paste0(path,"grid.png")) -->
<!-- ``` -->



<!-- ```{r, include=FALSE,eval =FALSE, warning=FALSE} -->
<!-- #pr =contour_plots(combined_data) -->
<!-- ########################################################################### -->
<!-- # plts1[[1]]|pr[["PRJNA453621"]][["contour_plot"]] -->
<!-- # plts1[[2]]|pr[["PRJEB45948"]][["contour_plot"]] -->
<!-- # plts1[[3]]|pr[["PRJNA589343"]][["contour_plot"]] -->
<!-- # plts1[[4]]|pr[["PRJNA687773"]][["contour_plot"]] -->
<!-- ####################################################################### -->
<!-- ``` -->





<!-- ```{r, echo=FALSE, include = FALSE, message=FALSE,fig.cap="\\label{figs2}  Plot of control against effect sizes for filtering threshold of 10,30 and 50. Data accession numbers: PRJNA453621, PRJEB45948, PRJNA589343 and PRJNA687773", fig.height = 6} -->
<!-- #  -->
<!-- # pr = Fit_control(logcontrol,plot_path,file_path, plotname) -->
<!-- # (pr[[1]]|pr[[2]])/(pr[[3]]|pr[[4]]) -->
<!-- # ############# -->
<!-- # deseq_data_list <- mget(load(paste0(path,"deseq_Filter_list.RData"))) -->
<!-- # deseq_data_list = deseq_data_list[[names(deseq_data_list)]] -->
<!-- # logcontrol_Filter_list <- mget(load(paste0(path,"logcontrol_Filter_list.RData"))) -->
<!-- # logcontrol_Filter_list <- logcontrol_Filter_list[[names(logcontrol_Filter_list)]] -->
<!-- #  -->
<!-- # v = Plot_ContEffects_multi_filters(deseq_Filter_list, -->
<!-- #                                    logcontrol_Filter_list) -->
<!-- # combined <- v[[1]] + v[[2]] + v[[3]] + v[[4]] & theme(legend.position = "bottom") -->
<!-- # combined + plot_layout(guides = "collect") -->


<!-- # effects_list <- list(data.frame(effect = deseq_list$PRJNA453621$log2FoldChange),  -->
<!-- #                            data.frame(effect = deseq_list$PRJEB45948$log2FoldChange), -->
<!-- #                            data.frame(effect = deseq_list$PRJNA589343$log2FoldChange), -->
<!-- #                            data.frame(effect =deseq_list$PRJNA687773$log2FoldChange))  -->
<!-- #  -->
<!-- # names(effects_list) <- names(deseq_list) -->
<!-- #  -->
<!-- # plotname  <- "effects_dist_" -->
<!-- # Fit_Effects(effects_list,logcontrol) -->
<!-- # param.controls  <- Control_Param[[names(Control_Param)]] -->
<!-- # param.effects = Effect_parameters[[names(Effect_parameters)]] -->

<!-- # param.effects <- mget(load(paste0(path,"Effect_Parameters.RData")))  -->
<!-- # param.effects <- param.effects[[names(param.effects)]] -->

<!-- # vv = HIST(effects_list,plot_path, plotname)[["hist"]] -->
<!-- # (vv[[1]]|vv[[2]])/(vv[[3]]|vv[[4]]) -->
<!-- ``` -->



<!-- <!-- My thoughts the contour plot for the low abundance otu. It is more less likely --> -->
<!-- <!-- that otus with low abundances have high effects than thoses with high abundance --> -->
<!-- <!-- so the left hand top quadrant should have less number of otus compared with --> -->
<!-- <!-- more number of otus found in the lower quadrant. Thus, the porportions of significant --> -->
<!-- <!-- otus found in the top leaft quadrant will mostly be lower compared to that of the top right --> -->
<!-- <!-- corner --> -->

# Conclusion
1.Fit control abundace, no show the plot for the power. That is bettwer
2. Show the plots for the power 






# References

<!-- \begin{table}[h] -->
<!-- \centering -->
<!-- \caption[Simple table]{Raw feaces samples} -->
<!-- \label{chap1Table1} -->
<!-- \begin{tabular}{p{2.5cm}|p{2.5cm}p{2.5cm}p{2.5cm}p{2.5cm}p{2.5cm}p{2.5cm}p{2.5cm}p{2.5cm}p{2.5cm}p{2.5cm}} % a vertical bar here will put one in the table -->
<!-- \toprule -->
<!-- Assession \\ number & Number of  samples &  Paired/single  &Platform  & Region &Primer & Link to  paper    m & m & m   \\ -->
<!-- \midrule -->
<!-- PRJNA687773 &  &  & & & & &  & & \\ -->
<!-- PRJNA624252 &  &  & & & & & & & \\ -->
<!-- PRJNA453621 &  &  & & & & & & & \\ -->
<!-- PRJNA642975 &  &  & & & & & & & \\ -->
<!-- PRJNA355023 &  &  & & & & & & & \\ -->
<!-- PRJNA168470 &  &  & & & & & & & \\ -->
<!-- PRJNA644763 &  &  & & & & & & & \\ -->
<!-- PRJNA578223 &  &  & & & & & & & \\ -->
<!-- PRJNA589343 &  &  & & & & & & & \\ -->
<!-- PRJEB45948  &  &  & & & & & & & \\  -->
<!-- \bottomrule -->
<!-- \end{tabular} -->
<!-- \end{table} -->
