---
title: "Power Analysis of Microbiome Data"
author: "Michael Agronah"
#date: "2022-10-27"
date: '\today'
#date: '`r format(Sys.time(), "%d %B, %Y")`'
output:
  pdf_document:
    toc: true    # can remove
    toc_depth: 3 # can remove
    number_sections: true
  highlight: zenburn
header-includes:
- \usepackage{fancyhdr}
- \usepackage{mathtools}
- \usepackage{xcolor, hyperref}
- \usepackage{lipsum}
- \usepackage{longtable}
- \usepackage{caption}
- \usepackage{diagbox}
- \usepackage{multirow}
- \usepackage{amsmath}
- \setlength{\headheight}{28pt}
- \setlength{\footskip}{25pt}
- \pagestyle{fancy}
- \usepackage{tabularx}
- \renewcommand{\headrulewidth}{0.5pt}
- \renewcommand{\footrulewidth}{0.5pt}
- \lhead{\includegraphics[width=8cm,height=1cm]{logo-McMaster}}
- \cfoot{School of Computational Science and Engineering \\ McMaster University}
- \rhead{\thepage}
- \hypersetup{colorlinks   = true, linkcolor=blue, urlcolor  = blue}
- \fancypagestyle{plain}{\pagestyle{fancy}}
editor_options:
  chunk_output_type: console
bibliography: packages.bib
---

```{r load_pkgs2, include=FALSE, echo=FALSE,message=FALSE}
library(ggplot2)
library(dplyr)
## uses denovo_simulation branch
library(glmmTMB)
library(MASS)
## remotes::install_github("chvlyl/ZIBR")
library(ZIBR)
## remotes::install_github("nyiuab/NBZIMM")
library(NBZIMM)
## library(splinectomeR)
library(lme4)
library(patchwork)
```


# Multivariate Analysis of Longitudinal Microbiome data using Generalised Linear Mixed Negative Binomial Model

## Abstract
Human microbiome is dynamic in nature. Understanding the dynamics in longitudinal microbiome data can help explain the mechanisms that underpin human health and diseases. However, longitudinal microbiome data analysis, whether 16S rRNA or metagenome shotgun, is difficult. Along with the difficulties presented by the characteristics of microbiome data such as sparsity, overdispersion, and compositionality, repeated measures introduce correlation between observations from the same individual at different time points. In an attempt to handle the problem of correlation, the majority of methods used in the literature to analyze longitudinal microbiome data only model one taxon at a single time point and do not account for the correlation structure between individual taxa across time.  In this project, a Multivariate Negative Binomial Mixed Model (MNBMM) is proposed for analyzing longitudinal microbiome data. A special feature of MNBMM is that it can handle the correlation structure of individual taxa over time. Fitting MNBMM to data, however, is very challenging because the number of parameters estimates for the unstructured covariance matrix grows quadratically with the dimensionality of the data. For atypical microbiome data with thousands of taxa, fitting this model is computationally expensive and almost impossible. A possible solution to implement a latent variable model for fitting a covariance matrix with fewer parameter estimates. In this project, a reduced rank method is used for fitting the data reduced rank method. The **glmmTMB** package in R provided a great functionality for fitting the reduced rank model. We demonstrate using from a simulation studies how implementation of the MNBMM works  in **glmmTMB**.   

<!-- In this project, Instead, a reduced rank method is used for fitting the data reduced rank method. We show from a simulation studies that the reduced rank model can fit the fixed effect parameters with lower error compared to the  -->

## Introduction
Longitudinal data arises when repeated measures are collected for the same subject. For example; DNA taking the microbiome of pregnant women repeatedly over multiple time points. The literature of longitudinal microbiome investigations has generally responded mainly to  two queries: (1). to determine how microbiome abundances change over time between groups (for example cases versus controls) [@lewis2015inflammation; @backhed2015dynamics], and 2. to investigate the relationship between microbiome abundances and other factors, for instance, environmental factors, clinical outcomes, etc) [@kodikara2022statistical; @chen2016two].

Special features of microbiome data such as variation in sequencing depth, high dimensionality, overdispersion, excess zeros and compositionality, poses challenges when
analyzing microbiome data[@xia2018statistical]. Variations in sequencing depth makes comparison of reads across samples difficult. The problem of overdispersion renders statistical methods that assume means to be less or equal to variance ineffective. For example, the Poisson model which assumes mean to be equal to variance will underestimate the variance in microbiome data. The high dimensionality of microbiome data leads to numerical
instabilities and convergence problems. Due to compositionality of microbiome data, statistical methods with normal distributional assumption cannot be used [@xia2018statistical].

Longitudinal microbiome data analysis poses additional challenges. First, repeated measurements exhibit correlation between observations taken from the same subject at various points. Longitudinal data also often tend to have more missing data from a person or data not taken from everyone at all the time period studied [@kodikara2022statistical]. For instance, a longitudinal study collecting fecal specimens from patients may have missing data due to patients dropping out of the study during the period of the specimen collection.

Several models such as the Zero-inflated beta regression Model, Negative binomial mixed model, Block Bootstrap Method, Zero-inflated Gaussian mixed models, Fast zero-inflated negative binomial mixed model and Bayesian semi-parametric generalized linear model have been proposed for the studies of longitudinal microbiome data [@kodikara2022statistical]. Many researchers have proposed a linear mixed model as a means to handle the correlation exhibited by taxa from repeated measures [@zhang2020zero].

In order to overcome the problem of correlations between observations from  individuals across time, the majority of longitudinal microbiome research analyzes only one taxa at a time [@chen2016two]. This approach, however, ignores the correlation structure between taxa from the individual subject. In this project, a Multivariate Negative Binomial Mixed Model (MNBMM) is proposed for the studying longitudinal microbiome data. MNBMM has the capacity to account for correlation instruction between individual taxa over time.


## Research Goals

The goal of this project is to

1. model the relationship between taxa abundances over time and covariate variables of interest (for instance, clinical and environmental factors such as age, groups or disease status of patients, among others), while accounting for the correlation structure between individual taxa within individual subjects. The unstructured correlation matrix, however, requires too many parameter estimates and is computationally infeasible for a typical ASV or OTU tables with thousands of taxa. Thus, a reduced rank latent variable modelâ€™s estimation of the structured correlation matrix will provide an understanding of the correlation between taxa on the community level.

2. model how microbiome abundances change over time between groups and determine taxa with differential abundances over time between groups.  


## Methods
### Model Description
The count data denoted by $Y_{ijt}$ is modeled by a negative binomial model defined as
\begin{align}
Y_{ijt} &\sim NB(mean = \mu_{ijt}, dispersion = \alpha_{it}),\\
\boldsymbol\mu &= g^{-1}(E) \nonumber \\
E &= X\beta + Zb,\,\,b \sim MVN(0, \Sigma) \nonumber
\end{align}
where $i,j$ and $t$ denote the $i^{th}$ taxa, $j^{th}$ sample and the $t^{th}$ time, respectively, and $Y_{i,j,t}$ denotes the count data for the $i^{th}$ taxa in the $j^{th}$ sample at time $t$. The function $g$ is the link function and $\boldsymbol\mu$ is a vector of means with entries $\mu_{ijt}$.  $X$ and $Z$ are the fixed and random effect model matrices respectively and $\beta$ is the fixed effect parameter. The vector $b$ is a multivariate deviate from a gaussian distribution with a variance covariance matrix $\Sigma$.


Given an OTU or ASV table of dimension $n$, the parameter estimates required   for $\Sigma$ are given by $n(n + 1)/2$. Thus, the number of parameter estimates grows quadratically with $n$. For an OTU table with 10 taxa for instance, $55(=10*11/2)$ parameter estimates will be required for $\Sigma$. Consequently, the typical microbiome data with thousands of taxa will require millions of parameter estimates, which is computationally infeasible. Additionally, the unstructured covariance matrix $\Sigma$ is often singular due  to linear independence of the columns. This results in numerical instability and convergence problems when fitting the model. A possible remedy is to use a latent variable model, to specify a low dimensional latent variable that describes the full correlation structure. This project applies a reduced rank latent variable model for estimating the structured covariance matrix. The reduced rank model expresses the columns of $\Sigma$ by a set of $r < n$ latent variables. The number of parameters required to be estimated for the reduced rank model with $r$ latent variables is given by $nr - \begin{pmatrix} r \\2 \end{pmatrix}$, hence, reducing the number of parameters to be estimated significantly to a linear function of $n$. The **glmmTMB** package [@brooks2017glmmtmb] in R provides a flexible framework for fitting mixed models with a range of response distributions and also has the capability to fit reduced rank variance structure. The **glmmTMB** package is used in this project for the estimation of the reduced rank model. 


```{r simulatedata2,echo=FALSE,message=FALSE,warning=FALSE }
nIndiv <- 20; nTime <- 10; nTaxa <-100
metadat <-function(nIndiv,nTime,nTaxa){
  dd <- expand.grid(subject = factor(1:nIndiv),
                  taxa =factor(1:nTaxa), time = (1:nTime))

  grp = (rep(factor(c("control", "treatment")),length.out=nIndiv))
  dd$group <- grp
  dd
}

metadata =metadat(nIndiv,nTime,nTaxa)

Theta_fun <- function(nTaxa){
  rowIndices <- rep(1:nTaxa, 1:nTaxa); colIndices <- sequence(1:nTaxa)
  template <- sparseMatrix(rowIndices, colIndices, x =
                           if_else(rowIndices==colIndices,1,0))
template@x
}
thet <- Theta_fun(nTaxa)


Beta <- c(1, 0.2, 0.1, 0.2)

sim <- function(dispersion,data=metadata,bet=Beta, Theta = thet,n=nIndiv){
 p  = simulate(~group*time+(-1 + taxa|subject),
              newdata = data,
              newparams = list(beta= bet, theta =Theta),
              family=negative.binomial(theta=dispersion))
 data$count = unlist(p)
 data
}

dd =sim(dispersion = 2)
training <- dd[dd$time < floor(nTime*0.8),]
test <- anti_join(dd, training, by='time')
```

## Application

### Data Simulation
Using the simulate function in **lme4**, we generated count data from a generalized linear mixed model with a negative binomial distribution. The simulation's parameters are displayed in Table \ref{tab:simPara}. The plot of the trajectory for three taxa from the simulated data is shown in Figure \ref{fig:explore}. A rank reduced rank random slope model with subjects as the grouping variable, taxa as the random effect variable, and groups and the interaction between group and time as the fixed effect variable was fitted to the data. The optimal dimension of 3 for the reduced rank was determined by comparing the AICs of the model fitted to dimensions ranging from 2 to 10. 
<!-- Table \ref{aic} displays the AIC values and dimensions. -->
```{r fitmodel1,include = TRUE, eval=FALSE, echo=TRUE,message=FALSE,warning=FALSE, fig.height = 3}
fit_list <- lapply(2:10,
                   function(d) {
                     fit.rr <-
                     glmmTMB(count ~ group*time+ rr(-1 + taxa|subject,d = d),
                                              data = dd, family = nbinom2)
                       })

# compare fits via AIC
aic_vec <- sapply(fit_list, AIC)
aic_vec - min(aic_vec, na.rm = TRUE)
```

\begin{table*}[h]
  \centering
  \caption{Parameter values used in data simulation}
    \begin{tabular}{l|cc}
          \multicolumn{1}{p{4.75em}}{\textbf{Effects}} & \multicolumn{1}{l}{\textbf{Other parameters}}  \\
          Intercept = 1 & Number of subjects = 20\\
          Group = 0.2   & Number of Taxa = 100 \\
          Time  =0.1   & Length of Time  = 10\\
          Group*Time = 0.2  &  Number of Groups = 2  \\
          Dispersion parameter = 2 & Random effect parameter ($\theta$) = Identity \\
    \end{tabular}%
  \label{tab:simPara}%
\end{table*}

```{r plot2, echo=FALSE,message=FALSE,warning=FALSE, fig.show='hold', fig.cap="\\label{fig:explore} Some simulated taxa profiles with time, group and group * time effects and 0.2 dispersion", fig.fullwidth=FALSE,fig.align = 'center', fig.height = 3}

p1<-dd[dd$taxa ==1,] %>%
    ggplot(aes(x=time, y=log(count),  colour= (group),
             group = (subject), shape= (group),
             linetype = (group)))+
    geom_line()+ geom_point()+ggtitle("Taxa 1")+ylab("Count (log scale)")+
    labs(y = "count (log scale)", linetype = "Group", shape = "Group", color="Group") +
    scale_color_manual(values=c("steelblue","darkviolet"))+
    theme(legend.position = "none")

p11<-dd[dd$taxa ==2,]  %>%
    ggplot(aes(x=time, y=log(count),  colour= (group),
             group = (subject), shape= (group),
             linetype = (group)))+
    geom_line()+ geom_point()+ggtitle("Taxa 2")+ylab("Count (log scale)")+
    labs(y = "count (log scale)", linetype = "Group", shape = "Group", color="Group") +
    scale_color_manual(values=c("steelblue","darkviolet"))+
    theme(legend.position = "none")

p111<-dd[dd$taxa ==3,]  %>%
    ggplot(aes(x=time, y=log(count),  colour= (group),
             group = (subject), shape= (group),
             linetype = (group)))+
    geom_line()+ geom_point()+ggtitle("Taxa 3")+ylab("Count (log scale)")+
    labs(y = "count (log scale)", linetype = "Group", shape = "Group", color="Group") +
    scale_color_manual(values=c("steelblue","darkviolet"))+
    theme()

p1|p11|p111
```

## Results
```{r error_compar2, eval= FALSE, include= FALSE, echo=FALSE,message=FALSE,warning=FALSE, }
   
  fit_glmmtmb_rr = glmmTMB(count ~ group*time+ rr(-1 + taxa|subject,3),
                              data = dd, family = nbinom2)


```

### Comparison with the Univariate Negative Binomial Mixed Model

The Univariate Negative Binomial Mixed Model (UNBMM)  has often been used in the literature of longitudinal microbiome studies [@kodikara2022statistical] . In this section we compare the effectiveness of UNBMM and MNBMM in predicting the fixed effect parameter. The UNBMM was fitted using the **glmm.nb** package developed by which is commonly used for longitudinal microbiome studies in the literature [@zhang2020nbzimm].  Figure \ref{errorplots} shows the mean square errors of the fixed effect parameter estimated from both models for 50 simulations. For simplicity, 10 taxa are used for this comparison study. The plot shows lower errors from the MNBMM model. Consequently, accounting for correlations between the individual taxa provides better estimates of the fixed effect parameters.

```{r error_compare, eval= TRUE, include= TRUE, echo=FALSE,message=FALSE,warning=FALSE, fig.cap="\\label{errorplots} Comparison of the mean squared error for the fixed effect estimates from the Univariate and Multivariate Negative Binomial Mixed Models"}

Compare_models <- function(nInd, nTax, Bet,nTim,nSim){
 
  Theta = Theta_fun(nTax);  Ncol = length(Bet)
  Other_Effect_Est <- matrix(NA, nrow = nTax, ncol = Ncol)
  other_err =  glmmtmb_err =  matrix(NA,  nrow = nSim,ncol = Ncol)

  for(i in 1:nSim){

    TrueEffects <- matrix(rep(Bet, each = nTax),
                          nrow = nTax,ncol = Ncol)

    metadata = metadat(nInd,nTim,nTax)
    
    dd = sim(dispersion=2,metadata,Bet,Theta,nInd)
    
    fit_glmmtmb_rr = glmmTMB(count ~ group*time+ rr(-1 + taxa|subject,3),
                              data = dd, family = nbinom2)
    
    glmmtmb_err[i,]  = (Bet - unlist(fixef(fit_glmmtmb_rr)$cond))^2

    for(n in 1:nTax){

      Taxa <- dd[dd$taxa == n,]
      fit_nbmm<-glmm.nb(count~group + time + group*time,
                        random = ~ 1 | subject, data = Taxa,
                        verbose = FALSE,
                        method = "ML", control =list(msMaxIter = 1000,
                                                     msMaxEval = 1000))

      Other_Effect_Est[n,] = fixef(fit_nbmm)
    }

    otherErr = (TrueEffects - Other_Effect_Est)^2
    other_err[i,] = colMeans(otherErr)
  }

  other_err_dat = data.frame(other_err)
  glmmtmb_err_dat = data.frame(glmmtmb_err)
  cols = c("intercept", "group", "time", "group_time")
  colnames(other_err_dat) = colnames(glmmtmb_err_dat) = cols

  v = list(other_err_data = other_err_dat, glmmtmb_err_data = glmmtmb_err_dat)

  nsim = 1:nSim
  inter <- (v
            %>% setNames(c("other", "glmmtmb"))
            %>% purrr::map_dfr(pull, intercept)
  )
  inter$nsim  = nsim
  grp <- (v
          %>% setNames(c("other", "glmmtmb"))
          %>% purrr::map_dfr(pull, group)
  )

  grp$nsim  = nsim
  tme <- (v
          %>% setNames(c("other", "glmmtmb"))
          %>% purrr::map_dfr(pull, time)
  )
  tme$nsim  = nsim

  grptime <- (v
              %>% setNames(c("other", "glmmtmb"))
              %>% purrr::map_dfr(pull, group_time)
  )
  grptime$nsim  = nsim
  v = list(Intercept = inter, Group  = grp, Time = tme, Group_Time = grptime)
}

nIndiv = 20; nTime = 10; nTaxa = 10; Beta  =  c(1, 0.2, 0.1, 0.2)
v = Compare_models(nInd=nIndiv, nTax=nTaxa,Bet=Beta,nTim =nTime,nSim=10)


effect_error_plots <- function(error_data_list){

  plts = list()
  for(i in 1:length(v)){
    plt = ggplot(v[[i]],aes(x =  nsim)) +
      geom_point(aes(y = other)) +
      geom_line(aes(y = other,color = "UNBMM")) +
      geom_point(aes( y = glmmtmb)) +
      geom_line(aes(y = glmmtmb,color = "MNBMM")) +
      ggtitle(names(v)[i]) +
      xlab("Number of simulations") + ylab("Mean Squared Errors")
      # ggtitle("Error comparision: Mutivariate ve univariate NB model")
    plts[[i]] = plt
  }
  plts
}

plts = effect_error_plots(v)
p = (plts[[1]]|plts[[2]])/(plts[[3]]|plts[[4]]) & theme(legend.position = "bottom")
p + plot_layout(guides = "collect")
```

### Model Prediction

This section shows the mean squared error of our model prediction for individual taxa for 50 simulations. The points on the plot in figure \ref{errorp} represent that average mean squared errors for individual taxa across time. Here, we consider 10 taxa. On average the mean squared error for these over taxa is approximately 4.96. 

```{r ,echo=FALSE,message=FALSE,warning=FALSE,out.width="50%",fig.cap="\\label{errorp} Average mean squared error for prediction of mean abundances of individual taxa across time"}

sim_sed <- function(dispersion,sed,data=metadata,bet=Beta, Theta = thet,n=nIndiv){
 p  = simulate(seed =sed, ~group*time+(-1 + taxa|subject:time),
              newdata = data,
              newparams = list(beta= bet, theta =Theta),
              family=negative.binomial(theta=dispersion))
 data$count = unlist(p)
 data
}

latent_vals <- function(sed){
   p = simulate(seed = sed, ~group*time + (-1 + taxa|subject),
              newdata = dd,
              newparams = list(beta = Beta,
                               theta = thet*1e16,
                               sigma = 1e-16),
              family = gaussian)[[1]]
  p
}

nIndiv <- 20; nTime <- 10; nTaxa <-10
metadata =metadat(nIndiv,nTime,nTaxa)
thet <- Theta_fun(nTaxa); dd = metadata
Beta <- c(1, 0.2, 0.1, 0.2)

nSim = 50; error_by_taxa = counts=matrix(NA, nrow = nrow(metadata),ncol = nSim)
for(i in 1:nSim){
  sed = i
  dat = sim_sed(dispersion=2,sed,data=metadata,bet=Beta, Theta = thet,n=nIndiv)
  latent = latent_vals(sed)
  glmmTMBfit_rr <- glmmTMB(count ~ group*time+ rr(-1 + taxa|subject, 2),
                   data = dat, family = nbinom2)
  pred = predict(glmmTMBfit_rr)

  X = model.matrix(~group*time,metadata)
  true_data = X%*%(Beta) + latent
  error_by_taxa[, i] = (true_data - pred)^2
}

MSE = rowMeans(error_by_taxa)
metadata$MSE = MSE

M = matrix(NA,nrow=nTime, ncol=nTaxa )
for(i in 1:nTime){
    data  = metadata[metadata$time== i,]
  for(j in 1:nTaxa){
    dat = data[data$taxa== j,]
    M[i,j] = mean(dat$MSE)
    }
}

y = colMeans(M); x = 1:length(y)
dat = data.frame(x,y)
ggplot(dat, aes(x=x,y=y))+
   geom_point() +
    geom_line() +
    xlab("Taxa") +  ylab("Average mean squared error") 

all=mean(y)
```

\newpage
## Conclusion

In this project, Multivariate Negative Binomial Mixed Model (MNBMM) was proposed for analyzing longitudinal microbiome data. The model is capable of handling correlations between taxa from individuals over time. The reduced rank component of MNBMM allows fitting of this model to microbiome data of large dimensions.  MNBMM also has the capability to handle overdispersion in microbiome data. This project demonstrates how to fit the model in the **glmmTMB** package in R using simulation studies.  We show using the simulation studies that the models provide better estimates of the fixed effect parameters compared to the traditional univariate Negative Binomial Mixed Model applied in the literature of longitudinal microbiome studies. 

## Further development 

A limitation of MNBMM is that it does not handle zero inflation found in microbiome data. Additionally, MNBMM cannot handle microbiome data presented in propositions and does not handle compositionality in microbiome data. Further work on this project will extend this model to include zero inflation component so as to account for the zero inflation seen in microbiome data. A new model with a framework similar to MNBMM, for example a Multivariate Gaussian Mixed Model will be developed to handle microbiome data presented as proportions. 

\newpage 

# References {-}

\setlength{\parindent}{-0.2in}
\setlength{\leftskip}{0.2in}
\noindent
<div id="refs"></div>
```{r refmgr references, results="asis", echo=FALSE}
# Print
```
\setlength{\parindent}{0in}
\setlength{\leftskip}{0in}
\setlength{\parskip}{0pt}